bioengine/engine.py

```diff
@@ -0,0 +1,128 @@
+"""The BioEngine class."""
+import inspect
+import sys
+import yaml
+import os
+import threading
+import uuid
+from types import ModuleType
+
+from bioengine.utils import dotdict
+from bioengine.exceptions import UnsupportedAPI
+
+
+class BioEngine:
+    """The BioEngine class."""
+
+    def __init__(self, vendor_api=None):
+        """Set up engine."""
+        self.services = []
+        self.packages = {}
+        self._service_lock = threading.Lock()
+        self.vendor_api = vendor_api
+
+    def generate_engine_api(self, package_info=None, make_api_module=True):
+        engine_api = dotdict(showMessage=print, register=self.register_service)
+        if self.vendor_api:
+            engine_api.update(self.vendor_api)
+        if package_info:
+            engine_api.getPackageInfo = lambda: package_info
+            engine_api.getConfig = lambda: package_info.get("config")
+
+        if make_api_module:
+            # make a fake module with api
+            mod = ModuleType("bioengine")
+            sys.modules[mod.__name__] = mod  # pylint: disable=no-member
+            mod.__file__ = mod.__name__ + ".py"  # pylint: disable=no-member
+            mod.api = engine_api
+        return engine_api
+
+    def register_service(self, **api):
+        """Set interface."""
+        if isinstance(api, dict):
+            api = {a: value for a, value in api.items() if not a.startswith("_")}
+        elif inspect.isclass(type(api)):
+            api = {a: getattr(api, a) for a in dir(api) if not a.startswith("_")}
+        else:
+            raise UnsupportedAPI
+        self.services.append(dotdict(**api))
+
+    def register(self, **kwarg):
+        """Register api resources."""
+        if kwarg["type"] == "service":
+            self.register_service(**kwarg)
+        else:
+            raise NotImplementedError
+
+    def execute(self, script):
+        """Execute a script via the api."""
+        package_info = {}
+        package_info["id"] = str(uuid.uuid4())
+        package_info["_locals"] = {}
+        self.generate_engine_api()
+        exec(script, package_info["_locals"])  # pylint: disable=exec-used
+        self.packages[package_info["id"]] = package_info
+
+    def load_package(self, package_dir):
+        """load a service package."""
+        with open(os.path.join(package_dir, "config.yaml"), "r") as f:
+            package_info = yaml.load(f, Loader=yaml.FullLoader)
+            package_info["id"] = str(uuid.uuid4())
+            package_info["_locals"] = {}
+            package_info["package_dir"] = package_dir
+            try:
+                self._service_lock.acquire()
+
+                entry_point = os.path.join(package_dir, package_info["entrypoint"])
```

Comment:
We can validate the `package_info` yaml config with `voluptuous`. That's a separate PR.